## Correction step 2/2  - predictionstring 교정 - 진행중

In [49]:
import re
import string
import os

import pandas as pd
import numpy as np
from tqdm import tqdm
from pathlib import Path

BASE_DIR = Path('../input/feedback-prize-2021')
TRAIN_DIR = BASE_DIR / 'train'



In [50]:
tr_df = pd.read_csv(BASE_DIR / 'correct_v1.csv')
# df = pd.read_csv(BASE_DIR + "train.csv")

tr_df.head(5)

id  discourse_id  discourse_start  discourse_end  \
0  423A1CA112E2  1.622628e+12              8.0          229.0   
1  423A1CA112E2  1.622628e+12            230.0          312.0   
2  423A1CA112E2  1.622628e+12            313.0          401.0   
3  423A1CA112E2  1.622628e+12            402.0          758.0   
4  423A1CA112E2  1.622628e+12            759.0          886.0   

                                      discourse_text discourse_type  \
0  Modern humans today are always on their phone....           Lead   
1  They are some really bad consequences when stu...       Position   
2  Some certain areas in the United States ban ph...       Evidence   
3  When people have phones, they know about certa...       Evidence   
4  Driving is one of the way how to get around. P...          Claim   

  discourse_type_num                                   predictionstring  \
0             Lead 1  1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 1...   
1         Position 1       45 46 47 48 49 50 51 52 53 54 55 56 57 58 59   
2         Evidence 1    60 61 62 63 64 65 66 67 68 69 70 71 72 73 74 75   
3         Evidence 2  76 77 78 79 80 81 82 83 84 85 86 87 88 89 90 9...   
4            Claim 1  139 140 141 142 143 144 145 146 147 148 149 15...   

   new_start  new_end                                 new_discourse_text  
0          8      229  Modern humans today are always on their phone....  
1        230      312  They are some really bad consequences when stu...  
2        313      400  Some certain areas in the United States ban ph...  
3        402      757  When people have phones, they know about certa...  
4        759      886  Driving is one of the way how to get around. P...

In [51]:
"""
correction #2 : cmp with discTxtf in train.csv and predStr word from train dir
"""

def get_tokens(ids):
    with open(TRAIN_DIR/f'{ids}.txt', 'r', encoding='UTF-8') as file: data = file.read()
    return data.split()

def get_text(ids):
    with open(TRAIN_DIR/f'{ids}.txt', 'r', encoding='UTF-8') as file: data = file.read()
    return data

def verify_text(row):
    tokens = get_tokens(row['id'])
    indices = [int(x) for x in row.predictionstring.split()]
    start = indices[0]
    end = indices[-1] + 1
    text = ' '.join(tokens[start:end])
    return text

verify_text(tr_df.loc[120722])



"handcart people who can't go school for study but they want to study,its give you education when you stay in home. people don't want to go school they can do distance learning. practically education which help to get success in our life. online and video conferencing help your education strong. you can ask any thing ,you can search anything which help to solve your problem ,which help to get you some knowledge. student"

In [52]:
# train_names, train_texts = [], []
# for f in tqdm(list(os.listdir(TRAIN_DIR))):
#     train_names.append(f.replace('.txt', ''))
#     train_texts.append(open(TRAIN_DIR/f, 'r', encoding='UTF-8').read())

# train_texts = pd.DataFrame({'id': train_names, 'text': train_texts})
# train_texts

tqdm.pandas()
tr_df['verify'] = tr_df.progress_apply(verify_text, axis=1)


100%|██████████| 144293/144293 [01:10<00:00, 2033.99it/s]


In [53]:
tr_df['correct'] = tr_df.progress_apply(lambda row: ' '.join(row.discourse_text.split()) == row.verify, axis=1)
tr_df.correct.value_counts()

100%|██████████| 144293/144293 [00:05<00:00, 24424.52it/s]


True     118223
False     26070
Name: correct, dtype: int64

In [55]:
for _, row in tr_df[tr_df.correct == False].sample(n=5).iterrows():
    print(row['id'])
    print(row.discourse_text)
    print(row.verify)

57076BE40C47
limiting the use of cars can be very good for our enviornment because it would make the air less polluted 
limiting the use of cars can be very good for our enviornment because it would make the air less polluted.
41CDAAA96888
Even though the driverless car may need help sometimes trying to get through traffic jams and it would need your assitance 
Even though the driverless car may need help sometimes trying to get through traffic jams and it would need your assitance,
EBA1B05B6624
So if just a little amount of help would do something. Even if you cleaned around your house would help. Doing dishes would even help 
So if just a little amount of help would do something. Even if you cleaned around your house would help. Doing dishes would even help.
D3C9713AEDC1
students may not be able to do these activities 
students may not be able to do these activities,
D0DCC58B6C2B
but we do it anyways so why not just let us do it in the open 
but we do it anyways so why not just let u

In [56]:
tr_df.head(10)

id  discourse_id  discourse_start  discourse_end  \
0  423A1CA112E2  1.622628e+12              8.0          229.0   
1  423A1CA112E2  1.622628e+12            230.0          312.0   
2  423A1CA112E2  1.622628e+12            313.0          401.0   
3  423A1CA112E2  1.622628e+12            402.0          758.0   
4  423A1CA112E2  1.622628e+12            759.0          886.0   
5  423A1CA112E2  1.622628e+12            887.0         1150.0   
6  423A1CA112E2  1.622628e+12           1151.0         1533.0   
7  423A1CA112E2  1.622628e+12           1534.0         1602.0   
8  423A1CA112E2  1.622628e+12           1603.0         1890.0   
9  423A1CA112E2  1.622628e+12           1891.0         2027.0   

                                      discourse_text        discourse_type  \
0  Modern humans today are always on their phone....                  Lead   
1  They are some really bad consequences when stu...              Position   
2  Some certain areas in the United States ban ph...              Evidence   
3  When people have phones, they know about certa...              Evidence   
4  Driving is one of the way how to get around. P...                 Claim   
5  That's why there's a thing that's called no te...              Evidence   
6  Sometimes on the news there is either an accid...              Evidence   
7  Phones are fine to use and it's also the best ...                 Claim   
8  If you go through a problem and you can't find...              Evidence   
9  The news always updated when people do somethi...  Concluding Statement   

       discourse_type_num                                   predictionstring  \
0                  Lead 1  1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 1...   
1              Position 1       45 46 47 48 49 50 51 52 53 54 55 56 57 58 59   
2              Evidence 1    60 61 62 63 64 65 66 67 68 69 70 71 72 73 74 75   
3              Evidence 2  76 77 78 79 80 81 82 83 84 85 86 87 88 89 90 9...   
4                 Claim 1  139 140 141 142 143 144 145 146 147 148 149 15...   
5              Evidence 3  163 164 165 166 167 168 169 170 171 172 173 17...   
6              Evidence 4  211 212 213 214 215 216 217 218 219 220 221 22...   
7                 Claim 2  282 283 284 285 286 287 288 289 290 291 292 29...   
8              Evidence 5  297 298 299 300 301 302 303 304 305 306 307 30...   
9  Concluding Statement 1  355 356 357 358 359 360 361 362 363 364 365 36...   

   new_start  new_end                                 new_discourse_text  \
0          8      229  Modern humans today are always on their phone....   
1        230      312  They are some really bad consequences when stu...   
2        313      400  Some certain areas in the United States ban ph...   
3        402      757  When people have phones, they know about certa...   
4        759      886  Driving is one of the way how to get around. P...   
5        887     1149  That's why there's a thing that's called no te...   
6       1151     1532  Sometimes on the news there is either an accid...   
7       1534     1602  Phones are fine to use and it's also the best ...   
8       1603     1890  If you go through a problem and you can't find...   
9       1891     2026  The news always updated when people do somethi...   

                                              verify  correct  
0  Modern humans today are always on their phone....     True  
1  They are some really bad consequences when stu...     True  
2  Some certain areas in the United States ban ph...     True  
3  When people have phones, they know about certa...     True  
4  Driving is one of the way how to get around. P...     True  
5  That's why there's a thing that's called no te...     True  
6  Sometimes on the news there is either an accid...     True  
7  Phones are fine to use and it's also the best ...     True  
8  If you go through a problem and you can't find...     True  
9  The news always updated when people do somethi...     True

discource_text와 predictionstring을 이용하여 개별 train 파일로부터 추출한 text 간 비교 => 26070